compiler/src/main/kotlin/io/github/aplcornell/viaduct/circuitcodegeneration/BackendCodeGenerator.kt

```diff
@@ -0,0 +1,352 @@
+package io.github.aplcornell.viaduct.circuitcodegeneration
+
+import com.squareup.kotlinpoet.AnnotationSpec
+import com.squareup.kotlinpoet.CodeBlock
+import com.squareup.kotlinpoet.FileSpec
+import com.squareup.kotlinpoet.FunSpec
+import com.squareup.kotlinpoet.INT
+import com.squareup.kotlinpoet.KModifier
+import com.squareup.kotlinpoet.MemberName
+import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
+import com.squareup.kotlinpoet.PropertySpec
+import com.squareup.kotlinpoet.SET
+import com.squareup.kotlinpoet.TypeName
+import com.squareup.kotlinpoet.TypeSpec
+import com.squareup.kotlinpoet.asClassName
+import com.squareup.kotlinpoet.joinToCode
+import io.github.aplcornell.viaduct.circuitanalysis.NameAnalysis
+import io.github.aplcornell.viaduct.runtime.Out
+import io.github.aplcornell.viaduct.runtime.ViaductGeneratedProgram
+import io.github.aplcornell.viaduct.runtime.ViaductRuntime
+import io.github.aplcornell.viaduct.syntax.Host
+import io.github.aplcornell.viaduct.syntax.circuit.CircuitCallNode
+import io.github.aplcornell.viaduct.syntax.circuit.CircuitDeclarationNode
+import io.github.aplcornell.viaduct.syntax.circuit.FunctionDeclarationNode
+import io.github.aplcornell.viaduct.syntax.circuit.InputNode
+import io.github.aplcornell.viaduct.syntax.circuit.LetNode
+import io.github.aplcornell.viaduct.syntax.circuit.OutputNode
+import io.github.aplcornell.viaduct.syntax.circuit.ProgramNode
+import io.github.aplcornell.viaduct.syntax.circuit.Variable
+import io.github.aplcornell.viaduct.syntax.circuit.VariableBindingNode
+import io.github.aplcornell.viaduct.syntax.types.ValueType
+import io.github.aplcornell.viaduct.util.FreshNameGenerator
+import java.util.LinkedList
+import java.util.Queue
+import javax.annotation.processing.Generated
+
+private class BackendCodeGenerator(
+    val program: ProgramNode,
+    val host: Host,
+    codeGenerator: (context: CodeGeneratorContext) -> CodeGenerator,
+    val hostDeclarations: TypeSpec,
+) {
+    private val nameAnalysis = NameAnalysis.get(program)
+
+    private val context = Context()
+    private val codeGenerator = codeGenerator(context)
+
+    fun generateClass(): TypeSpec {
+        val classBuilder = TypeSpec.classBuilder(host.name.replaceFirstChar { it.uppercase() }).primaryConstructor(
+            FunSpec.constructorBuilder().addParameter("runtime", ViaductRuntime::class).build(),
+        ).addProperty(
+            PropertySpec.builder("runtime", ViaductRuntime::class).initializer("runtime")
+                .addModifiers(KModifier.PRIVATE).build(),
+        )
+        // TODO Replace with a call to ProtocolAnalysis
+        val participatingProtocols = program.circuits.map { it.protocol.value }
```

Comment:
The OG compiler doesn't seem to have any particular ordering of protocols (the protocolanalysis returns a set). Should I do the same here, or order by alpha

Reply:
Since Kotlin sets remember insertion order, the OG compiler processes protocols in program order (this is very implicit and subtle, which is not great). Program order is an option (this is what I recommended for functions and circuits for example), but in this case I think sorting would be better. Since protocols aren't really top level and the programmer doesn't care about the order they appear in the program.

BTW, looking at circuit headers is not enough. Let statements in functions can have arbitrary protocols. Best way to do this is to traverse the program (using `.children`) and accumulating a set. See `Declarations.kt` for inspiration.